enterprise/coderd/coderd.go

```diff
@@ -450,11 +456,165 @@ func (api *API) updateEntitlements(ctx context.Context) error {
 		}
 	}
 
+	if changed, enabled := featureChanged(codersdk.FeatureWorkspaceProxy); changed {
+		if enabled {
+			fn := derpMapper(api.Logger, api.ProxyHealth)
+			api.AGPL.DERPMapper.Store(&fn)
+		} else {
+			api.AGPL.DERPMapper.Store(nil)
+		}
+	}
+
 	api.entitlements = entitlements
 
 	return nil
 }
 
+// getProxyDERPStartingRegionID returns the starting region ID that should be
+// used for workspace proxies. A proxy's actual region ID is the return value
+// from this function + it's RegionID field.
+//
+// Two ints are returned, the first is the starting region ID for proxies, and
+// the second is the maximum region ID that already exists in the DERP map.
+func getProxyDERPStartingRegionID(derpMap *tailcfg.DERPMap) (sID int, mID int) {
+	maxRegionID := 0
+	for _, region := range derpMap.Regions {
+		if region.RegionID > maxRegionID {
+			maxRegionID = region.RegionID
+		}
+	}
+	if maxRegionID < 0 {
+		maxRegionID = 0
+	}
+
+	// Round to the nearest 10,000 with a sufficient buffer of at least 2,000.
+	const roundStartingRegionID = 10_000
+	const startingRegionIDBuffer = 2_000
+	startingRegionID := maxRegionID + startingRegionIDBuffer
+	startingRegionID = int(math.Ceil(float64(startingRegionID)/roundStartingRegionID) * roundStartingRegionID)
+	if startingRegionID < roundStartingRegionID {
+		startingRegionID = roundStartingRegionID
+	}
+
+	return startingRegionID, maxRegionID
```

Comment:
This took me some time to understand. Can we add a comment?

```suggestion
	// Round to the nearest 10,000 with a sufficient buffer of at least 2,000.
	// The buffer is for .... (why do we need the buffer?)
	// Example: 
	//   maxRegionID = 8500
	//   startingRegionID = 10500
	//   bufferedStartID = 20000
	const roundStartingRegionID = 10_000
	const startingRegionIDBuffer = 2_000
	startingRegionID := maxRegionID + startingRegionIDBuffer
	bufferedStartID = int(math.Ceil(float64(startingRegionID)/roundStartingRegionID) * roundStartingRegionID)
	if bufferedStartID < roundStartingRegionID {
		bufferedStartID = roundStartingRegionID
	}

	return bufferedStartID, maxRegionID
```